File: dotnet/Families/GameX.Xbox/Formats/Binary.cs
using GameX.Uncore.Formats;
using GameX.Xbox.Formats.Xna;
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

namespace GameX.Xbox.Formats;

#region Binary_Xnb

public class Binary_Xnb : IHaveMetaInfo, IWriteToStream, Indirect<object> {
    public static Task<object> Factory(BinaryReader r, FileSource f, Archive s) => Task.FromResult((object)new Binary_Xnb(r, f));
    object Indirect<object>.Value => Obj;

    #region Headers

    const uint MAGIC = 0x00424e58; // XNB?

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct Header {
        public static (string, int) Struct = ("<I2bI", 10);
        public uint Magic;
        public byte Version;
        public byte Flags;
        public uint SizeOnDisk;
        public readonly bool Compressed => (Flags & 0x80) != 0;
        public readonly char Platform => (char)(Magic >> 24);

        public ContentReader Validate(BinaryReader r, string path) {
            if ((Magic & 0x00FFFFFF) != MAGIC) throw new Exception("BAD MAGIC");
            if (Version != 5 && Version != 4) throw new Exception("Invalid XNB version");
            if (SizeOnDisk > r.BaseStream.Length) throw new Exception("XNB file has been truncated.");
            if (Compressed) {
                uint decompressedSize = r.ReadUInt32(), compressedSize = SizeOnDisk - (uint)r.Tell();
                var b = r.DecompressXmem((int)compressedSize, (int)decompressedSize);
                return new ContentReader(new MemoryStream(b), path, Version, null);
            }
            return new ContentReader(r.BaseStream, path, Version, null);
        }
    }

    #endregion

    public object Obj;
    public bool AtEnd;

    public Binary_Xnb(BinaryReader r2, FileSource f) {
        var h = r2.ReadS<Header>();
        var r = h.Validate(r2, f.Path);
        Obj = r.ReadAsset<object>();
        AtEnd = r.AtEnd();
        //r.EnsureAtEnd(); // h.SizeOnDisk
    }

    public void WriteToStream(Stream stream) => Obj.Serialize(stream);

    public override string ToString() => Obj.Serialize();

    List<MetaInfo> IHaveMetaInfo.GetInfoNodes(MetaManager resource, FileSource file, object tag) => (Obj as IHaveMetaInfo)?.GetInfoNodes(resource, file, tag) ?? [
        new(null, new MetaContent { Type = "Data", Name = Path.GetFileName(file.Path), Value = this }),
        new("Xnb", items: [
            new($"Obj: {Obj}")
        ])];
}

#endregion

#region Binary_XXX

public unsafe class Binary_XXX : ArcBinary<Binary_XXX> {
    public override Task Read(BinaryArchive source, BinaryReader r, object tag) {
        var files = source.Files = [];
        return Task.CompletedTask;
    }

    public override Task<Stream> ReadData(BinaryArchive source, BinaryReader r, FileSource file, object option = default) {
        throw new NotImplementedException();
    }
}

#endregion

File: python/gamex/families/Xbox/formats/binary.py
from __future__ import annotations
import os, itertools
from io import BytesIO
from openstk import _throw, _pathExtension, Reader, IWriteToStream
from gamex import Archive, BinaryArchive, ArcBinary, ArcBinaryT, FileSource, MetaInfo, MetaManager, MetaContent, IHaveMetaInfo, DesSer
from gamex.families.Uncore.formats.compression import decompressXbox
from gamex.families.Xbox.formats.xna import ContentReader

# types
type Vector2 = ndarray
type Vector3 = ndarray
type Vector4 = ndarray
type Matrix4x4 = ndarray
type Quaternion = ndarray

#region Binary_Xnb

# Binary_Xnb
class Binary_Xnb(IHaveMetaInfo, IWriteToStream):
    @staticmethod
    def factory(r: Reader, f: FileSource, s: Archive): return Binary_Xnb(r, f)

    #region Headers

    MAGIC = 0x00424e58 #: XNB?

    class Header:
        _struct = ('<I2bI', 10)
        def __init__(self, tuple):
            self.magic, \
            self.version, \
            self.flags, \
            self.sizeOnDisk = tuple
        @property
        def compressed(self) -> bool: return (self.flags & 0x80) != 0
        @property
        def platform(self) -> chr: return chr(self.magic >> 24)

        def validate(self, r: Reader, path: str):
            if (self.magic & 0x00FFFFFF) != Binary_Xnb.MAGIC: raise Exception('BAD MAGIC')
            if self.version != 5 and self.version != 4: raise Exception('Invalid XNB version')
            if self.sizeOnDisk > r.f.getbuffer().nbytes: raise Exception('XNB file has been truncated.')
            if self.compressed:
                decompressedSize = r.readUInt32(); compressedSize = self.sizeOnDisk - r.tell()
                b = decompressXbox(r, compressedSize, decompressedSize)
                return ContentReader(BytesIO(b), path, self.version, None)
            return ContentReader(r.f, path, self.version, None)

    #endregion

    def __init__(self, r2: Reader, f: FileSource):
        h = r2.readS(self.Header)
        r = h.validate(r2, f.path)
        self.obj = r.readAsset()
        self.atEnd = r.atEnd()
        # r.ensureAtEnd()

    # @staticmethod
    # def addX(reader: Binary_Xnb.TypeReader) -> Binary_Xnb.TypeReader: pass
        
    def writeToStream(self, stream: object): return DesSer.serialize(self.obj, stream)

    def __repr__(self): return DesSer.serialize(self.obj)

    def getInfoNodes(self, resource: MetaManager = None, file: FileSource = None, tag: object = None) -> list[MetaInfo]: return self.obj.getInfoNodes(resource, file, tag) if isinstance(self.obj, IHaveMetaInfo) else [
        MetaInfo(None, MetaContent(type = 'Text', name = os.path.basename(file.path), value = self)),
        MetaInfo('Xnb', items = [
            MetaInfo(f'Obj: {self.obj}')
            ])
        ]


#endregion